File: Csharp/NyzoCL/NyzoTransaction.cs
using System;

namespace NyzoCL;

// Untested. Use at your own risk.
public class NyzoTransaction {
    public DateTime Timestamp {get;private set;}
    public short Type {get;private set;}
    public long Amount {get;private set;}
    public byte[] RecipientIdentifier {get;private set;}
    public long PreviousHashHeight {get;private set;}
    public byte[] PreviousBlockHash {get;private set;}
    public byte[] SenderIdentifier {get;private set;}
    public byte[] SenderData {get;private set;}
    public byte[] Signature {get;private set;}

    public NyzoTransaction(){
        this.Timestamp = DateTime.Now;
        this.Type = 2;
        this.Amount = 0L;
        this.RecipientIdentifier = new byte[32];
        this.PreviousHashHeight = 0;
        this.PreviousBlockHash = new byte[32];
        this.SenderIdentifier = new byte[32];
        this.SenderData = new byte[0];
        this.Signature = new byte[64];
    }

    public void SetTimestamp(DateTime timestamp){
        this.Timestamp = timestamp;
    }

    public void SetAmount(long amount){
        this.Amount = amount;
    }

    public void SetRecipientIdentifier(byte[] recipientIdentifier){
        for (var i = 0; i < 32; i++) {
            this.RecipientIdentifier[i] = recipientIdentifier[i];
        }
    }

    public void SetPreviousHashHeight(long previousHashHeight){
        this.PreviousHashHeight = previousHashHeight;
    }

    public void SetPreviousBlockHash(byte[] previousBlockHash){
        for (var i = 0; i < 32; i++) {
            this.PreviousBlockHash[i] = previousBlockHash[i];
        }
    }

    public void SetSenderIdentifier(byte[] senderIdentifier){
        for (var i = 0; i < 32; i++) {
            this.SenderIdentifier[i] = senderIdentifier[i];
        }
    }

    public void SetSenderData(byte[] senderData){
        this.SenderData = new byte[Math.Min(senderData.Length, 32)];
        for (var i = 0; i < this.SenderData.Length; i++) {
            this.SenderData[i] = senderData[i];
        }
    }

    public void SetSignature(byte[] signature){
        for (var i = 0; i < 64; i++) {
            this.Signature[i] = signature[i];
        }
    }

    public void Sign(byte[] seedBytes){
        var keyPair = Sodium.PublicKeyAuth.GenerateKeyPair(seedBytes);

        for (var i = 0; i < 32; i++) {
            this.SenderIdentifier[i] = keyPair.PublicKey[i];
        }

        var signature = Sodium.PublicKeyAuth.Sign(this.GetBytes(false)!, keyPair.PrivateKey);

        for (var i = 0; i < 64; i++) {
            this.Signature[i] = signature[i];
        }
    }

    public byte[]? GetBytes(bool includeSignature){
        var forSigning = !includeSignature;
        var buffer = new ByteBuffer(1000);

        buffer.PutByte(2); // transaction type = 2 (standard)
        buffer.PutLong(this.Timestamp.ToFileTimeUtc());
        buffer.PutLong(this.Amount);
        buffer.PutBytes(this.RecipientIdentifier);

        if(forSigning){
            buffer.PutBytes(this.PreviousBlockHash);
        } else {
            buffer.PutLong(this.PreviousHashHeight);
        }

        buffer.PutBytes(this.SenderIdentifier);

        if(forSigning){
            buffer.PutBytes(
                NyzoUtil.DoubleSha256(
                    this.SenderData
                )
            );
        } else {
            buffer.PutInt(this.SenderData.Length);
            buffer.PutBytes(this.SenderData);
        }

        if(!forSigning){
            buffer.PutBytes(this.Signature);
        }

        return buffer.ReadBytes();
    }

    public static NyzoTransaction FromBytes(byte[] array){
        var buffer = new ByteBuffer();
        buffer.PutBytes(array);

        var transaction = new NyzoTransaction();
        transaction.Type = buffer.ReadByte();
        transaction.SetTimestamp(DateTime.FromFileTimeUtc(buffer.ReadLong()));
        transaction.SetAmount(buffer.ReadLong());
        transaction.SetRecipientIdentifier(buffer.ReadBytes(32));
        transaction.SetPreviousHashHeight(buffer.ReadLong());
        transaction.SetSenderIdentifier(buffer.ReadBytes(32));

        var dataLength = buffer.ReadByte();
        transaction.SetSenderData(buffer.ReadBytes(dataLength));
        transaction.SetSignature(buffer.ReadBytes(64));

        return transaction;
    }
}

File: npm.Nyzo.CL/NyzoTransaction.js
const nacl = require("tweetnacl");

module.exports = class NyzoTransaction {
    constructor() {
        this.timestamp = Date.now();
        this.type = 2;
        this.amount = 0;
        this.recipientIdentifier = new Uint8Array(32);
        this.previousHashHeight = 0;
        this.previousBlockHash = new Uint8Array(32);
        this.senderIdentifier = new Uint8Array(32);
        this.senderData = new Uint8Array(0);
        this.signature = new Uint8Array(64);
    }

    SetTimestamp(timestamp) {
        this.timestamp = timestamp;
    }

    SetAmount(amount) {
        this.amount = amount;
    }

    SetRecipientIdentifier(recipientIdentifier) {
        for (let i = 0; i < 32; i++) {
            this.recipientIdentifier[i] = recipientIdentifier[i];
        }
    }

    SetPreviousHashHeight(previousHashHeight) {
        this.previousHashHeight = previousHashHeight;
    }

    SetPreviousBlockHash(previousBlockHash) {
        for (let i = 0; i < 32; i++) {
            this.previousBlockHash[i] = previousBlockHash[i];
        }
    }

    SetSenderIdentifier(senderIdentifier) {
        for (let i = 0; i < 32; i++) {
            this.senderIdentifier[i] = senderIdentifier[i];
        }
    }

    SetSenderData(senderData) {
        this.senderData = new Uint8Array(Math.min(senderData.length, 32));
        for (let i = 0; i < this.senderData.length; i++) {
            this.senderData[i] = senderData[i];
        }
    }

    SetSignature(signature) {
        for (let i = 0; i < 64; i++) {
            this.signature[i] = signature[i];
        }
    }

    Sign(seedBytes) {
        let keyPair = nacl.sign.keyPair.fromSeed(seedBytes);
        for (let i = 0; i < 32; i++) {
            this.senderIdentifier[i] = keyPair.publicKey[i];
        }

        let signature = nacl.sign.detached(this.GetBytes(false), keyPair.secretKey);
        for (let i = 0; i < 64; i++) {
            this.signature[i] = signature[i];
        }
    }

    GetBytes(includeSignature) {
        let forSigning = !includeSignature;
        let buffer = new ByteBuffer(1000);

        buffer.PutByte(2); // transaction type = 2 (standard)
        buffer.PutLong(this.timestamp);
        buffer.PutLong(this.amount);
        buffer.PutBytes(this.recipientIdentifier);

        if (forSigning) {
            buffer.PutBytes(this.previousBlockHash);
        } else {
            buffer.PutLong(this.previousHashHeight);
        }

        buffer.PutBytes(this.senderIdentifier);

        if (forSigning) {
            buffer.PutBytes(NyzoUtil.DoubleSha256(this.senderData));
        } else {
            buffer.PutByte(this.senderData.length);
            buffer.PutBytes(this.senderData);
        }

        if (!forSigning) {
            buffer.PutBytes(this.signature);
        }

        return buffer.ToArray();
    }

    static FromBytes(array) {
        const buffer = new ByteBuffer();
        buffer.InitializeWithArray(array);

        const transaction = new NyzoTransaction();
        transaction.type = buffer.ReadByte();
        transaction.SetTimestamp(buffer.ReadLong());
        transaction.SetAmount(buffer.ReadLong());
        transaction.SetRecipientIdentifier(buffer.ReadBytes(32));
        transaction.SetPreviousHashHeight(buffer.ReadLong());
        transaction.SetSenderIdentifier(buffer.ReadBytes(32));
        const dataLength = buffer.ReadByte();
        transaction.SetSenderData(buffer.ReadBytes(dataLength));
        transaction.SetSignature(buffer.ReadBytes(64));

        return transaction;
    }
}
